hydradx/model/amm/centralized_market.py

```diff
@@ -0,0 +1,254 @@
+                        tkns_bought += sell_tkns_remaining / ask[0]
+                        ask[1] -= sell_tkns_remaining / ask[0]
+                        sell_tkns_remaining = 0
+                    else:
+                        tkns_bought += ask[1]
+                        sell_tkns_remaining -= ask[1] * ask[0]
+                        ask[1] = 0
+                    if ask[1] == 0:
+                        remove_asks += 1
+                    if sell_tkns_remaining <= 0:
+                        break
+
+            agent.holdings[tkn_sell] -= sell_quantity - sell_tkns_remaining
+            agent.holdings[tkn_buy] += tkns_bought * (1 - self.trade_fee)
+
+        elif buy_quantity > 0:
+            buy_tkns_remaining = buy_quantity
+            tkns_sold = 0
+
+            if tkn_buy == base:
+                for ask in self.order_book[(base, quote)].asks:
+                    if ask[1] >= buy_tkns_remaining:
+                        tkns_sold += buy_tkns_remaining * ask[0]
+                        ask[1] -= buy_tkns_remaining
+                        buy_tkns_remaining = 0
+                    else:
+                        tkns_sold += ask[0] * ask[1]
+                        buy_tkns_remaining -= ask[1]
+                        ask[1] = 0
+                    if ask[1] == 0:
+                        remove_asks += 1
+                    if buy_tkns_remaining <= 0:
+                        break
+            else:
+                for bid in self.order_book[(base, quote)].bids:
+                    if bid[0] * bid[1] >= buy_tkns_remaining:
+                        tkns_sold += buy_tkns_remaining / bid[0]
+                        bid[1] -= buy_tkns_remaining / bid[0]
+                        buy_tkns_remaining = 0
+                    else:
+                        tkns_sold += bid[1]
+                        buy_tkns_remaining -= bid[0] * bid[1]
+                        bid[1] = 0
+                    if bid[1] == 0:
+                        remove_bids += 1
+                    if buy_tkns_remaining <= 0:
+                        break
+
+            agent.holdings[tkn_buy] += buy_quantity - buy_tkns_remaining
+            agent.holdings[tkn_sell] -= tkns_sold / (1 - self.trade_fee)
+            # remove these afterward, so we don't mess up the iteration
+            self.order_book[(base, quote)].bids = self.order_book[(base, quote)].bids[remove_bids:]
+            self.order_book[(base, quote)].asks = self.order_book[(base, quote)].asks[remove_asks:]
+
+        return self
+
+    def fail_transaction(self, error: str, **kwargs):
+        self.fail = error
+        return self
+
+    def copy(self):
+        return copy.deepcopy(self)
+
+    def buy_spot(self, tkn: str, numeraire: str = 'USD') -> float:
+        if tkn == numeraire:
+            return 1
+        elif (tkn, numeraire) in self.order_book:
+            return self.order_book[(tkn, numeraire)].asks[0][0] / (1 - self.trade_fee)
```

Comment:
I think this should be ask * (1 + trade_fee). It's not totally obvious from the Kraken documentation my they say they always denominate and charge the fee in the quote asset. The way I interpret this for a buy is that if you buy TKN with QTE, you pay 1% extra QTE.